src/deserializeSchema.ts

```diff
@@ -0,0 +1,485 @@
+    variants: [string, Fields][];
+};
+
+/**
+ * Map type.
+ */
+export type MapType = {
+    typeTag: ParameterType.Map;
+    sizeLength: SizeLength;
+    ofKeys: Type;
+    ofValues: Type;
+};
+
+/**
+ * List type.
+ */
+export type ListType = {
+    typeTag: ParameterType.List | ParameterType.Set;
+    sizeLength: SizeLength;
+    of: Type;
+};
+
+/**
+ * Pair type.
+ */
+export type PairType = {
+    typeTag: ParameterType.Pair;
+    ofLeft: Type;
+    ofRight: Type;
+};
+
+/**
+ * Reads {@link Type} from the given {@link Readable}.
+ *
+ * @param source input stream
+ * @returns contract schema type
+ */
+export function deserialType(source: Readable): Type {
+    const tag = deserialUint8(source);
+    switch (tag) {
+        case ParameterType.Unit:
+        case ParameterType.Bool:
+        case ParameterType.U8:
+        case ParameterType.U16:
+        case ParameterType.U32:
+        case ParameterType.U64:
+        case ParameterType.U128:
+        case ParameterType.I8:
+        case ParameterType.I16:
+        case ParameterType.I32:
+        case ParameterType.I64:
+        case ParameterType.I128:
+        case ParameterType.Amount:
+        case ParameterType.AccountAddress:
+        case ParameterType.ContractAddress:
+        case ParameterType.Timestamp:
+        case ParameterType.Duration:
+            return { typeTag: tag };
+        case ParameterType.Pair:
+            return {
+                typeTag: tag,
+                ofLeft: deserialType(source),
+                ofRight: deserialType(source),
+            };
+        case ParameterType.List:
+        case ParameterType.Set:
+            return {
+                typeTag: tag,
+                sizeLength: deserialUint8(source),
+                of: deserialType(source),
+            };
+        case ParameterType.Map:
+            return {
+                typeTag: tag,
+                sizeLength: deserialUint8(source),
+                ofKeys: deserialType(source),
+                ofValues: deserialType(source),
+            };
+        case ParameterType.Array:
+            return {
+                typeTag: tag,
+                size: deserialUint32(source),
+                of: deserialType(source),
+            };
+        case ParameterType.Struct:
+            return {
+                typeTag: tag,
+                fields: deserialFields(source),
+            };
+        case ParameterType.Enum:
+            return {
+                typeTag: tag,
+                variants: deserialArrayFn<[string, Fields]>(
+                    deserialTupleFn<string, Fields>(
+                        deserialString,
+                        deserialFields
+                    )
+                )(source),
+            };
+        case ParameterType.String:
+        case ParameterType.ContractName:
+        case ParameterType.ReceiveName:
+            return {
+                typeTag: tag,
+                sizeLength: deserialUint8(source),
```

Comment:
If String, ContractName and ReceiveName is always 32 bit, why do you read the length here?

Reply:
@shjortConcordium  thanks for the update. we have fixed this now.